### ISSUE TITLE
`@dlp.log_init` should support named parameters

### ISSUE THREAD

**Issue by rayandrew** [hariharan-devarajan/dlio-profiler#80]
Hi Hari! I am Ray, we've met before for Stormer I/O profiler.

Just want to let you know, while using this library, I tried to patch `FashionMNIST` dataset to incorporate `dlio` as I/O profiler.

This is the code that I used

```python
class CustomFashionMNIST(torchvision.datasets.FashionMNIST):
    @dlp.log_init
    def __init__(
        self,
        root,
        train=True,
        transform=None,
        target_transform=None,
        download=False,
    ):
        super().__init__(
            root=root,
            train=train,
            transform=transform,
            target_transform=target_transform,
            download=download,
        )

...

training_set = CustomFashionMNIST(
    "./data", train=True, transform=transform, download=True
)
```

However, I got an error:

```
Traceback (most recent call last):
  File "/lus/grand/projects/ReForMerS/rayandrew/stormer-dev/test_mnist.py", line 131, in <module>
    training_set = CustomFashionMNIST(
                   ^^^^^^^^^^^^^^^^^^^
TypeError: CustomFashionMNIST.__init__() got an unexpected keyword argument 'train'
```

I suspected this because `dlp.log_init` does not support named parameters.
I am actually able to fix it by patching the `log_init` function to this:

```diff
--- ./.venv/lib/python3.11/site-packages/dlio_profiler/logger.py	2024-05-10 22:39:22.000000000 +0000
+++ t.py	2024-05-10 22:39:20.000000000 +0000
@@ -229,14 +229,15 @@
             arg_names = inspect.getfullargspec(init)[0]
 
         @wraps(init)
-        def new_init(args, *kwargs):
+        def new_init(*args, **kwargs):
             if DLIO_PROFILER_ENABLE:
-                for name, value in zip(arg_names[1:], kwargs):
-                    setattr(args, name, value)
-                    if name == "epoch":
-                        self._arguments["epoch"] = str(value)
+                arg_values = dict(zip(arg_names[1:], args))
+                arg_values.update(kwargs)
+                if "epoch" in arg_values:
+                    arg_values["epoch"] = str(arg_values["epoch"])
+                self._arguments = arg_values
                 start = dlio_logger.get_instance().get_time()
-            init(args, *kwargs)
+            init(*args, **kwargs)
             if DLIO_PROFILER_ENABLE:
                 end = dlio_logger.get_instance().get_time()
```

Let me know if this actually works for you as well, thanks!

**Comment by hariharan-devarajan** [comment]
@rayandrew could u send a PR for this with appropriate test in the test.py so that I can merge.

**Comment by rayandrew** [comment]
Hi @hariharan-devarajan, I have opened a PR here #81

**Comment by hariharan-devarajan** [comment]
Closing this as it was fixed in #81
